src/nupic/regions/TMRegion.cpp

```diff
@@ -0,0 +1,855 @@
+    NTA_DEBUG << "compute " << *out << std::endl;
+  }
+  out = getOutput("predictedActiveCells");
+  if (out && (out->hasOutgoingLinks() || LogItem::isDebug())) {
+    tm_->getWinnerCells(*out->getData().getSDR());
+    NTA_DEBUG << "compute " << *out << std::endl;
+  }
+}
+
+
+/********************************************************************/
+
+Spec *TMRegion::createSpec() {
+  auto ns = new Spec;
+
+  ns->description =
+      "TMRegion. Class implementing the temporal memory algorithm as "
+      "described in 'BAMI "
+      "<https://numenta.com/biological-and-machine-intelligence/>'.  "
+      "The implementation here attempts to closely match the pseudocode in "
+      "the documentation. This implementation does contain several additional "
+      "bells and whistles such as a column confidence measure.";
+
+
+  /* ---- parameters ------ */
+
+  /* constructor arguments */
+  ns->parameters.add(
+      "numberOfCols",
+      ParameterSpec("(int) Number of mini-columns in the region. This values "
+                    "needs to be the same as the number of columns in the "
+                    "input from SP.  Normally this value is derived from "
+                    "the input width but if privided, this parameter must be "
+                    "the same total size as the input.",
+                    NTA_BasicType_UInt32,          // type
+                    1,                             // elementCount
+                    "",                            // constraints
+                    "0",                           // defaultValue
+                    ParameterSpec::CreateAccess)); // access
+
+  ns->parameters.add(
+      "cellsPerColumn",
+      ParameterSpec("(int) The number of cells per mini-column.",
+                    NTA_BasicType_UInt32, // type
+                    1,                    // elementCount
+                    "",                   // constraints
+                    "32",                 // defaultValue
+                    ParameterSpec::CreateAccess)); // access
+
+  ns->parameters.add(
+      "activationThreshold",
+      ParameterSpec("(int) Number of synapses that must be active to "
+                    "activate a segment.",
+                    NTA_BasicType_UInt32,          // type
+                    1,                             // elementCount
+                    "",                            // constraints
+                    "13",                          // defaultValue
+                    ParameterSpec::CreateAccess)); // access
+
+  ns->parameters.add(
+      "initialPermanence",
+      ParameterSpec("(float) Initial permanence for newly created synapses.",
+                    NTA_BasicType_Real32,          // type
+                    1,                             // elementCount
+                    "",                            // constraints
+                    "0.21",                        // defaultValue
+                    ParameterSpec::CreateAccess)); // access
```

Comment:
these are the params I call duplicit, at least shared with TM, BackTM (but I think there's a common ground for more regions (numColumns, permanenceInc/Dec,... ) 

>  /* constructor arguments */
  ns->parameters.add(
      "numberOfCols",
      ParameterSpec("(int) Number of mini-columns in the region. This values "
                    "needs to be the same as the number of columns in the "
                    "input from SP.  Normally this value is derived from "
                    "the input width but if privided, this parameter must be "
                    "the same total size as the input.",
                    NTA_BasicType_UInt32,          // type
                    1,                             // elementCount
                    "",                            // constraints
                    "0",                           // defaultValue
                    ParameterSpec::CreateAccess)); // access

  ns->parameters.add(
      "cellsPerColumn",
      ParameterSpec("(int) The number of cells per mini-column.",
                    NTA_BasicType_UInt32, // type
                    1,                    // elementCount
                    "",                   // constraints
                    "32",                 // defaultValue
                    ParameterSpec::CreateAccess)); // access

  ns->parameters.add(
      "activationThreshold",
      ParameterSpec("(int) Number of synapses that must be active to "
                    "activate a segment.",
                    NTA_BasicType_UInt32,          // type
                    1,                             // elementCount
                    "",                            // constraints
                    "13",                          // defaultValue
                    ParameterSpec::CreateAccess)); // access

  ns->parameters.add(
      "initialPermanence",
      ParameterSpec("(float) Initial permanence for newly created synapses.",
                    NTA_BasicType_Real32,          // type
                    1,                             // elementCount
                    "",                            // constraints
                    "0.21",                        // defaultValue
                    ParameterSpec::CreateAccess)); // access

  ns->parameters.add(
      "connectedPermanence",
      ParameterSpec("(float) ",
                    NTA_BasicType_Real32, // type
                    1,                    // elementCount
                    "",                   // constraints
                    "0.5",                // defaultValue
                    ParameterSpec::CreateAccess)); // access

Reply:
> these are the params I call duplicit, 

But remember, this is a plugin.  Each plugin must have its own configuration per the plugin framework which defines the behavior of that plugin.  The whole idea of a plugin is that each is independent with a clean interface that is uniform across all plugins.  

Besides, if you look carefully only numberOfCols and cellsPerColumn fields are really common but even  these have different defaults.  So they cannot be shared.  One alternative would be to change the plugin framework and use a text file containing this configuration information if having code that looks similar bothers you.

No, I think it is better that each plugin has its own set of functions as required by the plugin definition including its Spec and we always know where to find it.  If we change a spec parameter we know that only that plugin's behavior is changed.